lib/nolibc/stdio.c

```diff
@@ -541,3 +544,153 @@ int puts(const char *s)
 {
 	return fputs_internal(s, stdout, 1);
 }
+
+#if CONFIG_LIBVFSCORE
+struct _nolibc_file {
+	int fd;
+	bool eof;
+};
+
+/* The following code is derived from musl libc */
+static int __fmodeflags(const char *mode, int *flags)
+{
+	if (!strchr("rwa", *mode)) {
+		errno = EINVAL;
+		return -1;
+	}
+	if (strchr(mode, '+'))
+		*flags = O_RDWR;
+	else if (*mode == 'r')
+		*flags = O_RDONLY;
+	else
+		*flags = O_WRONLY;
+	if (strchr(mode, 'x'))
+		*flags |= O_EXCL;
+	if (strchr(mode, 'e'))
+		*flags |= O_CLOEXEC;
+	if (*mode != 'r')
+		*flags |= O_CREAT;
+	if (*mode == 'w')
+		*flags |= O_TRUNC;
+	if (*mode == 'a')
+		*flags |= O_APPEND;
+
+	return 0;
+}
+
+FILE *fopen(const char *pathname, const char *mode)
+{
+	int flags;
+	int fd;
+	int ret;
+
+	ret = __fmodeflags(mode, &flags);
+
+	if (ret < 0)
+		return NULL;
+
+	fd = open(pathname, flags, 0666);
+
+	if (fd < 0)
+		return NULL;
+
+	return fdopen(fd, mode);
+}
+
+int feof(FILE *stream)
+{
+	return stream->eof;
+}
+
+int fclose(FILE *stream)
+{
+	int ret = close(stream->fd);
+
+	free(stream);
+	return ret;
+}
+
+FILE *fdopen(int fd, const char *mode __unused)
+{
+	FILE *f = (FILE *)malloc(sizeof(FILE));
+
+	if (!f)
+		return NULL;
+	f->fd = fd;
+	f->eof = 0;
+	return f;
+}
+
+int fseek(FILE *stream, long offset, int whence)
+{
+	return lseek(stream->fd, offset, whence);
+}
+
+size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream)
+{
+	if (unlikely(!stream || !ptr || !size || !nmemb)) {
+		errno = EINVAL;
+		return 0;
+	}
+
+	if (unlikely(stream->fd < 0)) {
+		errno = EBADF; // Bad file descriptor
+		return 0;
+	}
+
+	if (unlikely(SIZE_MAX / size < nmemb)) {
+		errno = EOVERFLOW;
+		return 0;
+	}
+
+	size_t total = 0;
+
+	while (total < size * nmemb) {
+		ssize_t ret = read(stream->fd,
```

Comment:
Hum, I am not sure if it is maybe safer to use `pread()` (when `fd` supports it) and store `seek`  by our own. A program may malfunction due to an incorrect `seek` i fa program `reads` by itself form the same `fd` while it also interacts with it via `fdopen()`. Do you know how `musl` solves this case?

Obviously, this comment applies also for `fwrite()`.